test/unit/local/steps/buffer.js

```diff
@@ -0,0 +1,372 @@
+
+/*::
+import type {
+  AtomWatcherEvent,
+  Batch
+} from '../../../../core/local/steps/event'
+*/
+
+const builders = new Builders()
+
+function nonEmptyBatch (batchNumber /*: number */ = 1) /*: AtomWatcherEvent[] */ {
+  return [
+    builders.event().action('created').kind('file').path(`file-from-batch-${batchNumber}`).build(),
+    builders.event().action('deleted').kind('directory').path(`dir-from-batch-${batchNumber}`).build()
+  ]
+}
+
+should.Assertion.add('pending', function () {
+  this.params = {operator: 'be pending'}
+  this.obj.isPending().should.be.true()
+})
+
+describe('core/local/steps/Buffer', function () {
+  this.timeout(100)
+
+  describe('Basics', () => {
+    let buffer
+
+    beforeEach('instanciate', () => {
+      buffer = new Buffer()
+    })
+
+    describe('#push()', () => {
+      it('inserts the given Batch into the Buffer so it may be retrieved with #pop()', async () => {
+        const batch = nonEmptyBatch()
+
+        buffer.push(batch)
+
+        await should(buffer.pop()).be.fulfilledWith(batch)
+      })
+
+      it('inserts the same batch multiple times', async () => {
+        const batch = nonEmptyBatch()
+
+        buffer.push(batch)
+        buffer.push(batch)
+
+        await should(buffer.pop()).be.fulfilledWith(batch)
+        await should(buffer.pop()).be.fulfilledWith(batch)
+      })
+
+      it('inserts an empty batch', async () => {
+        const batch = []
+
+        buffer.push(batch)
+
+        await should(buffer.pop()).be.fulfilledWith(batch)
+      })
+    })
+
+    describe('#pop()', () => {
+      it('successively resolves with each Batch in insertion order', async () => {
+        const batch1 = nonEmptyBatch(1)
+        const batch2 = nonEmptyBatch(2)
+
+        buffer.push(batch1)
+        buffer.push(batch2)
+
+        await should(buffer.pop()).be.fulfilledWith(batch1)
+        await should(buffer.pop()).be.fulfilledWith(batch2)
+      })
+
+      describe('when all Batches were already popped', () => {
+        beforeEach('push and pop', async () => {
+          buffer.push(nonEmptyBatch(1))
+          await buffer.pop()
+        })
+
+        it('awaits for the next Batch', async () => {
+          const batch2 = nonEmptyBatch(2)
+
+          const nextBatchPromise = buffer.pop()
+          buffer.push(batch2)
+
+          await should(nextBatchPromise).be.fulfilledWith(batch2)
+        })
+      })
+
+      describe('when Buffer was never pushed anything', () => {
+        it('awaits for the next Batch', async () => {
+          const batch1 = nonEmptyBatch(1)
+          const batch2 = nonEmptyBatch(2)
+
+          const nextBatchPromise = buffer.pop()
+          buffer.push(batch1)
+          buffer.push(batch2)
+
+          await should(nextBatchPromise).be.fulfilledWith(batch1)
+        })
+      })
+
+      describe('when called more than once on an empty Buffer (DO NOT DO THIS)', () => {
+        it('only resolves the last Promise with the next Batch', async () => {
```

Comment:
Isn't this a bug?

Reply:
Well, I wonder :)
Ideally we should either prevent it in the implementation, or support it.

Reply:
Also this may not be our most important issue to fix since we don't have the need to do this for now.